File: TopInterview150/Hashmap/HappyNumber/solution.py
class Solution(object):
    
    def sumOfSquares(self, n):
        sum = 0
        while (n > 0):
            digit = n%10
            sum += digit*digit
            n//=10
        return sum
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        set = {-1}
        while n > 1 and n not in set:
            set.add(n)
            n = self.sumOfSquares(n)
        return n==1
    

File: TopInterview150/Hashmap/HappyNumber/solution.cpp
class Solution {
public:
    bool isHappy(int n) {
        unordered_set<int> set;
        while (n > 1 && set.find(n) == set.end())
        {
            set.insert(n);
            n = sumOfSquares(n);
        }
        return n == 1;
        
    }
    int sumOfSquares(int n)
    {
        int sum = 0;
        while (n > 0)
        {
            sum += (n%10)*(n%10);
            n = n/10;
        }
        return sum;
    }
};